File: Valestrom/Utils/src/net/verdagon/vale/Range.scala
package net.verdagon.vale


object CodeLocationS {
  // Keep in sync with CodeLocation2
  val testZero = CodeLocationS.internal(-1)
  def internal(internalNum: Int): CodeLocationS = {
    vassert(internalNum < 0)
    CodeLocationS(FileCoordinate("", Vector.empty, "internal"), internalNum)
  }
}

object RangeS {
  // Should only be used in tests.
  val testZero = RangeS(CodeLocationS.testZero, CodeLocationS.testZero)

  def internal(internalNum: Int): RangeS = {
    vassert(internalNum < 0)
    RangeS(CodeLocationS.internal(internalNum), CodeLocationS.internal(internalNum))
  }
}

case class CodeLocationS(
  // The index in the original source code files list.
  // If negative, it means it came from some internal non-file code.
  file: FileCoordinate,
  offset: Int) {
  val hash = runtime.ScalaRunTime._hashCode(this); override def hashCode(): Int = hash;

  // Just for debug purposes
  override def toString: String = {
    if (file == FileCoordinate.test) {
      "tv" + ":" + offset
    } else {
      file.toString + ":" + offset
    }
  }
}

case class RangeS(begin: CodeLocationS, end: CodeLocationS) {
  val hash = runtime.ScalaRunTime._hashCode(this); override def hashCode(): Int = hash;
  vassert(begin.file == end.file)
  vassert(begin.offset <= end.offset)
  def file: FileCoordinate = begin.file

  // Just for debug purposes
  override def toString: String = {
    if (file == FileCoordinate.test) {
      "tv" + ":" + begin.offset + "-" + end.offset
    } else {
      begin.toString + "-" + end.toString
    }
  }
}

File: FrontendRust/src/utils/range.rs
use crate::utils::code_hierarchy::FileCoordinate;
use crate::scout_arena::ScoutArena;

impl<'a> RangeS<'a> {
  pub fn internal(scout_arena: &ScoutArena<'a>, internal_num: i32) -> RangeS<'a> {
    assert!(internal_num < 0, "RangeS::internal - internal_num must be negative");
    RangeS::new(
      CodeLocationS::internal(scout_arena, internal_num),
      CodeLocationS::internal(scout_arena, internal_num))
  }
  
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
pub struct CodeLocationS<'a> {
  pub file: &'a FileCoordinate<'a>,
  pub offset: i32,
}

impl<'a> CodeLocationS<'a> {
  // Keep in sync with CodeLocation2
  pub fn test_zero(scout_arena: &ScoutArena<'a>) -> CodeLocationS<'a> {
    Self::internal(scout_arena, -1)
  }

  // SPORK
  pub fn internal(scout_arena: &ScoutArena<'a>, internal_num: i32) -> CodeLocationS<'a> {
    assert!(internal_num < 0, "CodeLocationS::internal - internal_num must be negative");
    let package_coord =
      scout_arena.intern_package_coordinate(scout_arena.intern_str(""), &[]);
    let file = scout_arena.intern_file_coordinate(package_coord, "internal");
    CodeLocationS {
      file,
      offset: internal_num,
    }
  }
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
pub struct RangeS<'a> {
  pub begin: CodeLocationS<'a>,
  pub end: CodeLocationS<'a>,
}

// Scala's toString was just for debug purposes (covered by #[derive(Debug)])
impl<'a> RangeS<'a> {
  pub fn new(begin: CodeLocationS<'a>, end: CodeLocationS<'a>) -> RangeS<'a> {
    assert!(begin.file == end.file, "RangeS: begin.file != end.file");
    assert!(begin.offset <= end.offset, "RangeS: begin.offset > end.offset");
    RangeS { begin, end }
  }

  // Should only be used in tests.
  pub fn test_zero(scout_arena: &ScoutArena<'a>) -> RangeS<'a> {
    let tz = CodeLocationS::test_zero(scout_arena);
    RangeS::new(tz.clone(), tz)
  }

  // SPORK
  pub fn file(&self) -> &'a FileCoordinate<'a> {
    self.begin.file
  }
}

